template/conftest.py

```diff
@@ -68,53 +74,66 @@ def pytest_generate_tests(metafunc):
-        if not fw_list:
+
+        # If "all" was specified, run on all firmwares
+        if metafunc.config.getoption("all"):
             for fw in FIRMWARES:
                 fw_list.append(fw)
                 ids.append(fw.device + " " + fw.version)
-        metafunc.parametrize("firmware", fw_list, ids=ids)
 
+        else:
+            # Enable only demanded firmwares
+            for fw in FIRMWARES:
+                if metafunc.config.getoption(fw.device):
+                    fw_list.append(fw)
+                    ids.append(fw.device + " " + fw.version)
 
-def prepare_speculos_args(firmware: Firmware, display: bool):
+        if not fw_list:
+            raise ValueError(f"No device specified")
+
+        metafunc.parametrize("firmware", fw_list, ids=ids, scope="session")
+
+
+def prepare_speculos_args(firmware: Firmware, display: bool, elfs_dir: str):
     speculos_args = []
 
     if display:
         speculos_args += ["--display", "qt"]
 
-    app_path = app_path_from_app_name(APPS_DIRECTORY, APP_NAME, firmware.device)
+    if elfs_dir == "bin":
+        app_path = APP_ROOT_DIR / elfs_dir / "app.elf"
+        assert app_path.is_file(), f"{app_path} must exist"
+    else:
+        app_path = app_path_from_app_name(APP_ROOT_DIR / elfs_dir, APP_NAME, firmware.device)
 
     return ([app_path], {"args": speculos_args})
 
 
 # Depending on the "--backend" option value, a different backend is
 # instantiated, and the tests will either run on Speculos or on a physical
 # device depending on the backend
-def create_backend(backend_name: str, firmware: Firmware, display: bool, log_apdu_file: Optional[Path]):
+def create_backend(backend_name: str, firmware: Firmware, display: bool, elfs_dir: str, log_apdu_file: Optional[Path]):
     if backend_name.lower() == "ledgercomm":
         return LedgerCommBackend(firmware=firmware, interface="hid", log_apdu_file=log_apdu_file)
     elif backend_name.lower() == "ledgerwallet":
         return LedgerWalletBackend(firmware=firmware, log_apdu_file=log_apdu_file)
     elif backend_name.lower() == "speculos":
-        args, kwargs = prepare_speculos_args(firmware, display)
+        args, kwargs = prepare_speculos_args(firmware, display, elfs_dir)
         return SpeculosBackend(*args, firmware=firmware, log_apdu_file=log_apdu_file, **kwargs)
     else:
         raise ValueError(f"Backend '{backend_name}' is unknown. Valid backends are: {BACKENDS}")
 
 
-# This final fixture will return the properly configured backend, to be used in tests
-@pytest.fixture
-def backend(backend_name, firmware, display, log_apdu_file):
-    with create_backend(backend_name, firmware, display, log_apdu_file) as b:
+# This fixture will return the properly configured backend, to be used in tests.
+# If your tests needs to be run on independent Speculos instances (in case they affect
+# setting for example), then you should change this fixture scope.
+@pytest.fixture(scope="session")
```

Comment:
Can't this already be done by using test classes that instantiate the backend as a class fixture?
If it is the case we should document it and try it on some of the boilerplate tests, this would allow a finetuning of the backend scope

Otherwise it would be nice to be able to have multiple backend scopes: per file / per test / for everyone / etc